// locklift/internal/tracing/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTracing = exports.Tracing = void 0;
const tracingInternal_1 = require("./tracingInternal");
class Tracing {
    constructor(ever, tracingInternal, features, network) {
        this.ever = ever;
        this.tracingInternal = tracingInternal;
        this.features = features;
        this.network = network;
        this.trace = async (transactionProm, config) => {
            const finalizedTx = await this.features.waitFinalized(transactionProm);
            return this.tracingInternal
                .trace({ finalizedTx, ...config })
                .then(traceTree => ({ ...finalizedTx.extTransaction, traceTree: traceTree }));
        };
        this.setAllowedCodes = (...params) => this.tracingInternal.setAllowedCodes(...params);
        this.setAllowedCodesForAddress = (...params) => this.tracingInternal.setAllowedCodesForAddress(...params);
        this.removeAllowedCodesForAddress = (...params) => this.tracingInternal.removeAllowedCodesForAddress(...params);
        this.removeAllowedCodes = (...params) => this.tracingInternal.removeAllowedCodes(...params);
        this.setContractLabels = tracingInternal.setContractLabels;
    }
    get allowedCodes() {
        return this.tracingInternal.allowedCodes;
    }
}
exports.Tracing = Tracing;
const createTracing = ({ ever, factory, features, tracingTransport, network, }) => {
    const internalTracing = new tracingInternal_1.TracingInternal(ever, factory, tracingTransport, network);
    return new Tracing(ever, internalTracing, features, network);
};
exports.createTracing = createTracing;
