// everscale-standalone-client/client/ConnectionController/proto.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProtoSocket = void 0;
const core_1 = __importDefault(require("../../core"));
const { nekoton, fetch, fetchAgent } = core_1.default;
class ProtoSocket {
    async connect(params) {
        class ProtoSender {
            constructor(params) {
                this.endpoint = params.endpoint;
                this.endpointAgent = fetchAgent(this.endpoint);
            }
            send(data, handler, _) {
                (async () => {
                    try {
                        const response = await fetch(this.endpoint, {
                            method: 'post',
                            headers: DEFAULT_HEADERS,
                            body: new Uint8Array(data),
                            agent: this.endpointAgent,
                        }).then(response => response.arrayBuffer());
                        handler.onReceive(new Uint8Array(response));
                    }
                    catch (e) {
                        handler.onError(e);
                    }
                })();
            }
        }
        return new nekoton.ProtoConnection(new ProtoSender(params));
    }
}
exports.ProtoSocket = ProtoSocket;
const DEFAULT_HEADERS = {
    'Content-Type': 'application/x-protobuf',
};
